ouroboros-network-framework/src/Simulation/Network/Snocket.hs

```diff
@@ -780,10 +780,30 @@ mkSnocket state tr = Snocket { getLocalAddr
                                    (Just remoteAddress)
                                    (STBearerInfo bearerInfo))
             -- connection delay
-            unmask (threadDelay (biConnectionDelay bearerInfo `min` connectTimeout))
-                   `onException`
-                   atomically (modifyTVar (nsConnections state)
-                                          (Map.delete $ normaliseId connId))
+            --
+            -- We need a way for a node to detect if the other end failed so
+            -- we keep an eye on the network state while waiting the full amount
+            -- of connection delay
+            -- TODO: Improve this see #3628
+            connDelayTimeoutVar <-
+              registerDelay (biConnectionDelay bearerInfo `min` connectTimeout)
+            unmask
+              (atomically $ runFirstToFinish $
+                  FirstToFinish (do
+                    LazySTM.readTVar connDelayTimeoutVar >>= check
+                    )
+                  <>
+                  FirstToFinish (do
+                    b <- not . Map.member (normaliseId connId)
+                      <$> readTVar (nsConnections state)
+                    check b
```

Comment:
Yes! https://github.com/input-output-hk/ouroboros-network/issues/3627 explains the particular error that happens. We need a way to detect failures like this while on `threadDelay`, otherwise we get invalid transitions. Note that this issue won't happen in real life since TCP will take care of it.